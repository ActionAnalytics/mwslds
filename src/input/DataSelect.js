import React from 'react'
import PropTypes from 'prop-types'

import Select from './Select'
import withData from '../datafetching/DataLoader'
import withToken from '../datafetching/Token'

const propTypes = {
  data: PropTypes.arrayOf(PropTypes.string),
  loading: PropTypes.bool,
  error: PropTypes.object,
}

const defaultProps = {
  data: null,
  loading: null,
  error: null,
}

function DataSelect(props) {
  const {
    loading,
    data,
    error,
    ...otherProps
  } = props

  if (loading) {
    return <Select data={['', 'loading...']} {...otherProps} />
  } else if (error) {
    return <Select data={['Error Fetching Data']} {...otherProps} disabled />
  }
  // expect prop type warning hnyaa, data[0] object instead of string
  const parsedData = [''].concat(data.map(val => val.code))
  return <Select data={parsedData} {...otherProps} />
}

DataSelect.propTypes = propTypes
DataSelect.defaultProps = defaultProps

export default withToken(withData(DataSelect))
