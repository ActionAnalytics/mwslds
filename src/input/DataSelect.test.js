import React from 'react'
import { configure, shallow, mount } from 'enzyme'
import { createSerializer } from 'enzyme-to-json'
import Adapter from 'enzyme-adapter-react-16'

import { UnwrappedDataSelect } from './DataSelect'

configure({ adapter: new Adapter() })
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }))

describe('DataSelect', () => {
  describe('UnwrappedDataSelect', () => {
    it('renders', () => {
      const select = mount(<UnwrappedDataSelect onChange={() => {}} />)
      expect(select).toMatchSnapshot()
    })

    it('is disabled when loading', () => {
      const select = mount(<UnwrappedDataSelect loading onChange={() => {}} />)
      expect(select).toMatchSnapshot()
      expect(select.find('Select').props().disabled).toBe(true)
    })

    it('is disabled on error', () => {
      const select = mount(<UnwrappedDataSelect error={{ error: 'oops' }} onChange={() => {}} />)
      expect(select).toMatchSnapshot()
      expect(select.find('Select').props().disabled).toBe(true)
    })

    it('appends empty data', () => {
      const data = [
        { value: 'a', label: 'A' },
      ]
      const select = mount(<UnwrappedDataSelect data={data} onChange={() => {}} />)
      expect(select).toMatchSnapshot()
      expect(select.find('Select').props().data[0]).toMatchObject({ value: '', label: '' })
    })

    it('updates when data changes', () => {
      const props = {
        data: [
          { value: 'a', label: 'A' },
        ],
        value: '',
        loading: false,
        validationError: null,
      }

      const select = shallow(<UnwrappedDataSelect {...props} onChange={() => {}} />)
      // same props, should not update
      expect(select.instance().shouldComponentUpdate(props)).toBe(false)

      const newData = [
        { value: 'b', label: 'B' },
      ]
      const newProps = Object.assign({}, props, { data: newData })
      // new data, should always update
      expect(select.instance().shouldComponentUpdate(newProps)).toBe(true)
    })
  })
})
