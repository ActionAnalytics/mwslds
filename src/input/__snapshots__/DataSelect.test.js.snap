// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSelect UnwrappedDataSelect appends empty data 1`] = `
<label
  className="control-label"
  htmlFor=""
  style={
    Object {
      "width": null,
    }
  }
>
  <div
    className="Select is-clearable is-searchable Select--single"
  >
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="Select-multi-value-wrapper"
        id="react-select-5--value"
      >
        <div
          className="Select-placeholder"
        >
          Select...
        </div>
        <div
          className="Select-input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-activedescendant="react-select-5--value"
            aria-expanded="false"
            aria-haspopup="false"
            aria-owns=""
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "5px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          />
        </div>
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
</label>
`;

exports[`DataSelect UnwrappedDataSelect is disabled on error 1`] = `
<label
  className="control-label"
  htmlFor=""
  style={
    Object {
      "width": null,
    }
  }
>
  <div
    className="Select is-clearable is-disabled is-searchable Select--single"
  >
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="Select-multi-value-wrapper"
        id="react-select-4--value"
      >
        <div
          className="Select-placeholder"
        >
          Select...
        </div>
        <div
          aria-activedescendant="react-select-4--value"
          aria-disabled="true"
          aria-expanded={false}
          aria-owns=""
          className="Select-input"
          onBlur={[Function]}
          onFocus={[Function]}
          role="combobox"
          style={
            Object {
              "border": 0,
              "display": "inline-block",
              "width": 1,
            }
          }
          tabIndex={0}
        />
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
</label>
`;

exports[`DataSelect UnwrappedDataSelect is disabled when loading 1`] = `
<label
  className="control-label"
  htmlFor=""
  style={
    Object {
      "width": null,
    }
  }
>
  <div
    className="Select is-clearable is-disabled is-searchable Select--single"
  >
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="Select-multi-value-wrapper"
        id="react-select-3--value"
      >
        <div
          className="Select-placeholder"
        >
          Select...
        </div>
        <div
          aria-activedescendant="react-select-3--value"
          aria-disabled="true"
          aria-expanded={false}
          aria-owns=""
          className="Select-input"
          onBlur={[Function]}
          onFocus={[Function]}
          role="combobox"
          style={
            Object {
              "border": 0,
              "display": "inline-block",
              "width": 1,
            }
          }
          tabIndex={0}
        />
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
</label>
`;

exports[`DataSelect UnwrappedDataSelect renders 1`] = `
<label
  className="control-label"
  htmlFor=""
  style={
    Object {
      "width": null,
    }
  }
>
  <div
    className="Select is-clearable is-searchable Select--single"
  >
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="Select-multi-value-wrapper"
        id="react-select-2--value"
      >
        <div
          className="Select-placeholder"
        >
          Select...
        </div>
        <div
          className="Select-input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-activedescendant="react-select-2--value"
            aria-expanded="false"
            aria-haspopup="false"
            aria-owns=""
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "5px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          />
        </div>
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
</label>
`;

exports[`DataSelect WrappedDataSelect renders 1`] = `
<label
  className="control-label"
  htmlFor=""
  style={
    Object {
      "width": null,
    }
  }
>
  <div
    className="Select is-clearable is-searchable Select--single"
  >
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
    >
      <span
        className="Select-multi-value-wrapper"
        id="react-select-6--value"
      >
        <div
          className="Select-placeholder"
        >
          Select...
        </div>
        <div
          className="Select-input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-activedescendant="react-select-6--value"
            aria-expanded="false"
            aria-haspopup="false"
            aria-owns=""
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "5px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          />
        </div>
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}
      >
        <span
          className="Select-arrow"
          onMouseDown={[Function]}
        />
      </span>
    </div>
  </div>
</label>
`;
